lib/web_ui/lib/src/ui/path_metrics.dart

```diff
@@ -120,12 +142,375 @@ class PathMetric {
   // [Iterator.current]. In this case, the [PathMetric] is valid before
   // calling `_moveNext` - `_moveNext` should be called after the first
   // iteration is done instead of before.
-  bool _moveNext() => throw UnimplementedError();
+  bool _moveNext() {
+    if (_subPathIndex == (_path.subpaths.length - 1)) {
+      return false;
+    }
+    ++_subPathIndex;
+    _buildSegments();
+    return true;
+  }
+
+  void _buildSegments() {
+    _segments = <_PathSegment>[];
+    _isClosed = false;
+    double distance = 0.0;
+    int pointIndex = -1;
+    bool haveSeenMoveTo = false;
+    bool haveSeenClose = false;
+
+    if (_path.subpaths.isEmpty) {
+      _contourLength = 0;
+      return;
+    }
+    final engine.Subpath subpath = _path.subpaths[_subPathIndex];
+    final List<engine.PathCommand> commands = subpath.commands;
+    double currentX = 0.0, currentY = 0.0;
+    final Function lineToHandler = (double x, double y) {
+      final double dx = currentX - x;
+      final double dy = currentY - y;
+      final double prevDistance = distance;
+      distance += math.sqrt(dx * dx + dy * dy);
+      // As we accumulate distance, we have to check that the result of +=
+      // actually made it larger, since a very small delta might be > 0, but
+      // still have no effect on distance (if distance >>> delta).
+      if (distance > prevDistance) {
+        _segments.add(_PathSegment(engine.PathCommandTypes.lineTo, distance,
+            [currentX, currentY, x, y]));
+      }
+      currentX = x;
+      currentY = y;
+    };
+    for (engine.PathCommand command in commands) {
+      switch (command.type) {
+        case engine.PathCommandTypes.moveTo:
+          final engine.MoveTo moveTo = command;
+          currentX = moveTo.x;
+          currentY = moveTo.y;
+          haveSeenMoveTo = true;
+          break;
+        case engine.PathCommandTypes.lineTo:
+          assert(haveSeenMoveTo);
```

Comment:
What would guarantee that a moveTo would precede a lineTo? Is it possible for this to throw in cases where the C++ engine doesn't?

Reply:
Followed skia implementation, it throws as well.